File: platform/macos/Bridge/NativeRuntime.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import TidyVNC

@MainActor
public final class NativeRuntime {
  let handle: NativeHandle
  private struct WeakSession { weak var value: NativeSession? }
  private var sessions: [WeakSession] = []
  private struct WeakListener { weak var value: NativeListener? }
  private var listeners: [WeakListener] = []
  private var shutdownTask: Task<Void, Error>?
  public init(sessionCapacity: UInt32 = 16) throws {
    var options = abi(tidyvnc_runtime_options.self)
    try checked { tidyvnc_runtime_options_init(&options, $0) }
    options.session_capacity = sessionCapacity
    options.required_features = UInt64(TIDYVNC_FEATURE_RUNTIME | TIDYVNC_FEATURE_TCP_UNIX_CONNECT |
      TIDYVNC_FEATURE_EVENT_POLL | TIDYVNC_FEATURE_IMAGES | TIDYVNC_FEATURE_INPUT |
      TIDYVNC_FEATURE_PROMPTS | TIDYVNC_FEATURE_CALLBACKS | TIDYVNC_FEATURE_GEOMETRY | TIDYVNC_FEATURE_CLIPBOARD | TIDYVNC_FEATURE_ENCODING | TIDYVNC_FEATURE_ENDPOINT_VALIDATION | TIDYVNC_FEATURE_SCALING | TIDYVNC_FEATURE_TILE_RENDERER | TIDYVNC_FEATURE_DAMAGE_GEOMETRY | TIDYVNC_FEATURE_CURSOR_RENDERER | TIDYVNC_FEATURE_INPUT_POLICY | TIDYVNC_FEATURE_SHORTCUTS | TIDYVNC_FEATURE_INPUT_RELEASE | TIDYVNC_FEATURE_CONNECTION_INFO | TIDYVNC_FEATURE_ENDPOINT_IDENTITY | TIDYVNC_FEATURE_PROMPT_SECURITY | TIDYVNC_FEATURE_CERTIFICATE_POLICY | TIDYVNC_FEATURE_HOST_KEY_ENCODING | TIDYVNC_FEATURE_REQUIRED_TLS_FILES | TIDYVNC_FEATURE_SECURITY_SELECTION | TIDYVNC_FEATURE_TLS_PRIORITY_VALIDATION | TIDYVNC_FEATURE_SECURITY_RECONFIGURATION | TIDYVNC_FEATURE_SHARED_SESSION | TIDYVNC_FEATURE_DESKTOP_LAYOUT | TIDYVNC_FEATURE_DISPLAY_LAYOUT) | UInt64(TIDYVNC_FEATURE_CANVAS_GEOMETRY) | UInt64(TIDYVNC_FEATURE_INPUT_TIMING) | UInt64(TIDYVNC_FEATURE_MESSAGE_LIMITS) | UInt64(TIDYVNC_FEATURE_WINDOW_GEOMETRY)
    var raw: UInt64 = 0; try checked { tidyvnc_runtime_create(&options, &raw, $0) }
    handle = NativeHandle(adopting: raw)
  }
  public func makeSession(configuration: NativeSessionConfiguration = .init()) throws -> NativeSession {
    guard shutdownTask == nil else { throw NativeError(.closing, "Runtime is shutting down") }
    let session = try NativeSession(runtime: self, configuration: configuration)
    sessions.removeAll { $0.value == nil }; sessions.append(WeakSession(value: session)); return session
  }
  // A session owns its runtime; this registry is weak and cannot create a cycle.
  public func makeListener(options: NativeListenOptions = .init()) throws -> NativeListener {
    guard shutdownTask == nil else { throw NativeError(.closing,"Runtime is shutting down") }
    let listener = try NativeListener(runtime:self,options:options)
    listeners.removeAll { $0.value == nil }; listeners.append(WeakListener(value:listener)); return listener
  }
  // Begin closing every session before awaiting any one of them.
  public func shutdown() async throws {
    if let shutdownTask { return try await shutdownTask.value }
    let closing = sessions.compactMap(\.value)
    let tasks = listeners.compactMap(\.value).map { $0.beginClose() } + closing.map { $0.beginClose() }
    let owner = handle
    let task = Task {
      try checked { tidyvnc_runtime_shutdown(owner.raw, $0) }
      var failure: (any Error)?
      for task in tasks { do { try await task.value } catch { if failure == nil { failure = error } } }
      do { try await waitForNativeDrain(owner, .runtime) } catch { if failure == nil { failure = error } }
      if let failure { throw failure }
    }
    shutdownTask = task; try await task.value
  }
  deinit { _ = tidyvnc_runtime_shutdown(handle.raw, nil) }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeRuntime.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

/// <summary>
/// Owns one core runtime (NativeRuntime.swift). UI-thread object: sessions and
/// listeners register weakly; shutdown begins closing all of them before
/// awaiting any.
/// </summary>
public sealed class NativeRuntime
{
    /// <summary>Capabilities the Windows app relies on; runtime creation fails without them.</summary>
    public const ulong RequiredFeatures =
        Tidyvnc.TIDYVNC_FEATURE_RUNTIME | Tidyvnc.TIDYVNC_FEATURE_TCP_UNIX_CONNECT | Tidyvnc.TIDYVNC_FEATURE_EVENT_POLL |
        Tidyvnc.TIDYVNC_FEATURE_IMAGES | Tidyvnc.TIDYVNC_FEATURE_INPUT | Tidyvnc.TIDYVNC_FEATURE_PROMPTS |
        Tidyvnc.TIDYVNC_FEATURE_CALLBACKS | Tidyvnc.TIDYVNC_FEATURE_GEOMETRY | Tidyvnc.TIDYVNC_FEATURE_CLIPBOARD |
        Tidyvnc.TIDYVNC_FEATURE_ENCODING | Tidyvnc.TIDYVNC_FEATURE_ENDPOINT_VALIDATION | Tidyvnc.TIDYVNC_FEATURE_SCALING |
        Tidyvnc.TIDYVNC_FEATURE_TILE_RENDERER | Tidyvnc.TIDYVNC_FEATURE_DAMAGE_GEOMETRY | Tidyvnc.TIDYVNC_FEATURE_CURSOR_RENDERER |
        Tidyvnc.TIDYVNC_FEATURE_INPUT_POLICY | Tidyvnc.TIDYVNC_FEATURE_SHORTCUTS | Tidyvnc.TIDYVNC_FEATURE_INPUT_RELEASE |
        Tidyvnc.TIDYVNC_FEATURE_CONNECTION_INFO | Tidyvnc.TIDYVNC_FEATURE_ENDPOINT_IDENTITY | Tidyvnc.TIDYVNC_FEATURE_PROMPT_SECURITY |
        Tidyvnc.TIDYVNC_FEATURE_CERTIFICATE_POLICY | Tidyvnc.TIDYVNC_FEATURE_HOST_KEY_ENCODING |
        Tidyvnc.TIDYVNC_FEATURE_REQUIRED_TLS_FILES | Tidyvnc.TIDYVNC_FEATURE_SECURITY_SELECTION |
        Tidyvnc.TIDYVNC_FEATURE_TLS_PRIORITY_VALIDATION | Tidyvnc.TIDYVNC_FEATURE_SECURITY_RECONFIGURATION |
        Tidyvnc.TIDYVNC_FEATURE_SHARED_SESSION | Tidyvnc.TIDYVNC_FEATURE_DESKTOP_LAYOUT | Tidyvnc.TIDYVNC_FEATURE_DISPLAY_LAYOUT |
        Tidyvnc.TIDYVNC_FEATURE_CANVAS_GEOMETRY | Tidyvnc.TIDYVNC_FEATURE_INPUT_TIMING | Tidyvnc.TIDYVNC_FEATURE_MESSAGE_LIMITS |
        Tidyvnc.TIDYVNC_FEATURE_WINDOW_GEOMETRY | Tidyvnc.TIDYVNC_FEATURE_ROUTED_CONNECT;

    internal NativeHandle Handle { get; }
    public IUiDispatcher Dispatcher { get; }
    private readonly List<WeakReference<NativeSession>> sessions = [];
    private readonly List<WeakReference<NativeListener>> listeners = [];
    private Task? shutdownTask;

    public unsafe NativeRuntime(IUiDispatcher dispatcher, uint sessionCapacity = 16)
    {
        Dispatcher = dispatcher;
        var options = Abi.Init<tidyvnc_runtime_options>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_runtime_options_init(&options, &error), &error);
        options.session_capacity = sessionCapacity;
        options.required_features = RequiredFeatures;
        ulong raw = 0;
        Abi.Check(NativeMethods.tidyvnc_runtime_create(&options, &raw, &error), &error);
        Handle = NativeHandle.Adopt(raw);
    }

    public bool IsShuttingDown => shutdownTask is not null;

    public NativeSession CreateSession(NativeSessionConfiguration? configuration = null)
    {
        UiThread.Require(Dispatcher);
        if (shutdownTask is not null) throw new NativeError(NativeStatus.Closing, "Runtime is shutting down");
        var session = new NativeSession(this, configuration ?? new NativeSessionConfiguration());
        sessions.RemoveAll(entry => !entry.TryGetTarget(out _));
        sessions.Add(new WeakReference<NativeSession>(session));
        return session;
    }

    public NativeListener CreateListener(NativeListenOptions? options = null)
    {
        UiThread.Require(Dispatcher);
        if (shutdownTask is not null) throw new NativeError(NativeStatus.Closing, "Runtime is shutting down");
        var listener = new NativeListener(this, options ?? new NativeListenOptions());
        listeners.RemoveAll(entry => !entry.TryGetTarget(out _));
        listeners.Add(new WeakReference<NativeListener>(listener));
        return listener;
    }

    /// <summary>Begins closing every listener and session, then awaits them and the runtime drain.</summary>
    public Task ShutdownAsync()
    {
        UiThread.Require(Dispatcher);
        return shutdownTask ??= Shutdown();
    }

    private async Task Shutdown()
    {
        var closing = new List<Task>();
        foreach (var entry in listeners) if (entry.TryGetTarget(out var listener)) closing.Add(listener.BeginClose());
        foreach (var entry in sessions) if (entry.TryGetTarget(out var session)) closing.Add(session.BeginClose());
        StartShutdown();
        Exception? failure = null;
        foreach (var task in closing)
        {
            try { await task.ConfigureAwait(true); } catch (Exception e) { failure ??= e; }
        }
        try { await NativeDrainer.WaitAsync(Handle, NativeDrain.Runtime).ConfigureAwait(true); } catch (Exception e) { failure ??= e; }
        if (failure is not null) throw failure;
    }

    private unsafe void StartShutdown()
    {
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_runtime_shutdown(Handle.Raw, &error), &error);
    }

    ~NativeRuntime()
    {
        unsafe { if (!Handle.IsClosed) _ = NativeMethods.tidyvnc_runtime_shutdown(Handle.Raw, null); }
    }

    /// <summary>The loaded core's ABI description (tidyvnc_get_abi).</summary>
    public static unsafe NativeAbiInfo GetAbi()
    {
        var info = Abi.Init<tidyvnc_abi_info>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_get_abi(&info, &error), &error);
        var types = new uint[Math.Min(info.security_count, 32u)];
        for (var i = 0; i < types.Length; i++) types[i] = info.security_types[i];
        return new NativeAbiInfo(info.features, info.handle_capacity, info.runtime_capacity, types);
    }
}

public sealed record NativeAbiInfo(ulong Features, uint HandleCapacity, uint RuntimeCapacity, IReadOnlyList<uint> SecurityTypes)
{
    public bool Supports(ulong feature) => (Features & feature) == feature;
}
